src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
     public void add(T value, int index) {
+        validateIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            addToHead(newNode);
+        } else {
+            addByIndex(newNode, index);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> returnNode = getNodeByIndex(index);
+        T returnValue = returnNode.item;
+        returnNode.item = value;
+        return returnValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        unlink(nodeToRemove);
+        size--;
+        T returnValue = nodeToRemove.item;
+        nodeToRemove.item = null;
+        return returnValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if ((object == current.item)
+                    || (object != null && object.equals(current.item))) {
+                unlink(current);
+                current.item = null;
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlink(Node<T> nodeToUnlink) {
+        if (nodeToUnlink.equals(head)) {
+            head = head.next;
+            nodeToUnlink.next = null;
+        } else if (nodeToUnlink.equals(tail)) {
+            tail = tail.prev;
+            nodeToUnlink.prev = null;
+        } else {
+            nodeToUnlink.prev.next = nodeToUnlink.next;
+            nodeToUnlink.next.prev = nodeToUnlink.prev;
+        }
+    }
+
+    private void addToHead(Node<T> node) {
+        if (head == null) {
+            head = node;
+            tail = node;
+        } else {
+            head.prev = node;
+            node.next = head;
+            head = node;
+        }
+    }
+
+    private void addByIndex(Node<T> node, int index) {
+        if (index == size) {
+            tail.next = node;
+            node.prev = tail;
+            tail = node;
+        } else {
+            Node<T> nodeByIndex = getNodeByIndex(index);
+            node.prev = nodeByIndex.prev;
+            nodeByIndex.prev.next = node;
+            nodeByIndex.prev = node;
+            node.next = nodeByIndex;
+        }
+    }
+
+    private void throwException() {
+        throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+    }
```

Comment:
i guess we can remove that method and throw new Exception when it appear

Reply:
It shortens the methods it's being used in. I think it looks cute =)

Reply:
but it also does not give us information on what is this exception/message and reduces readability + a method 1 statement long is not always, but usually a not-that-great idea :)